internal/objectstore/remote/retriever.go

```diff
@@ -234,54 +231,52 @@ func newRetriever(index int, remote apiremotecaller.RemoteConnection, newObjectC
 	return t
 }
 
-// Retrieve requests a blob from the remote API server, if there is not
-// remote connection, it will retry a number of times before returning an
-// error. This is allow time for the API connection to come up on a new HA node,
-// or after a restart.
-// If the blob isn't found or any other non-retryable error it will return an
-// error right away.
-// If the context is cancelled, it will stop processing the request as soon as
-// possible.
+// Retrieve requests a blob from the remote API server, if there is not a remote
+// connection, it will retry a number of times before returning an error. This
+// is to allow time for the API connection to come up on a new HA node, or after
+// a restart. If the blob isn't found or any other non-retryable error it will
+// return an error right away. If the context is cancelled, it will stop
+// processing the request as soon as possible.
 func (t *retriever) Retrieve(ctx context.Context, namespace, sha256 string) (io.ReadCloser, int64, error) {
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	select {
-	case <-t.tomb.Dying():
-		return nil, -1, tomb.ErrDying
-	case <-ctx.Done():
-		return nil, -1, ctx.Err()
-	case conn, ok := <-t.remote.Connection(ctx):
-		if !ok {
-			return nil, -1, NoRemoteConnection
-		}
-
+	var (
+		reader io.ReadCloser
+		size   int64
+	)
+	err := t.remote.Connection(ctx, func(ctx context.Context, conn api.Connection) error {
 		go func() {
 			defer cancel()
 
+			// If the connection is broken, we want to stop processing the
+			// request as soon as possible.
 			select {
 			case <-t.tomb.Dying():
+			case <-ctx.Done():
 			case <-conn.Broken():
```

Comment:
With the new design, I feel that the `conn.Broken()` -> cancelling the context now belongs in the apiremotecaller `Connection` method.

Reply:
Then this becomes nice:

```go
func (t *retriever) Retrieve(ctx context.Context, namespace, sha256 string) (io.ReadCloser, int64, error) {
	ctx := t.tomb.Context(ctx)
	var (
		reader io.ReadCloser
		size   int64
	)
	err := t.remote.Connection(ctx, func(ctx context.Context, conn api.Connection) error {
		httpClient, err := conn.RootHTTPClient()
		// ...
		reader, size, err = client.GetObject(ctx, namespace, sha256)
		return err
	})
	return reader, size, err
}
```

Reply:
Agreed.

Done.